### PR TITLE
Add printf '%#F' to output floats in hexadecimal OCaml constants

### PR DESCRIPTION
**Pull request by proux01**
Previous (bad) idea was to modify %h:

Make %h output valid OCaml constants

This was already the case except for:
* neg_infinity which was printed as "-infinity"
* some NaNs which were printed as "-nan"

Now, neg_infinity is printed as "neg_infinity" and all NaNs are
printed as "nan".

Scanf and float_of_string are expanded accordingly.

BTW, for those who'd like to know, we stepped on this while adding floating-point arithmetic
to the native computation scheme of Coq with @erikmd (coq/coq#9867).

### CODE REVIEW

**Comment by dbuenzli** [comment]
I don't think this is a good idea. The current format is compatible with the format read by [`strod`](http://pubs.opengroup.org/onlinepubs/000095399/functions/strtod.html). `%h` is useful for precise floating point data interchange, not for printing out OCaml constants.

Following established format OCaml convention for strings, char and floats, namely that uppercase format string means "OCaml syntax", I think we should rather only change the format for `%H` (hoping no-one is using it for interchange at the moment).

**Comment by proux01** [comment]
> I don't think this is a good idea. The current format is compatible with the format read by [`strod`](http://pubs.opengroup.org/onlinepubs/000095399/functions/strtod.html). `%h` is useful for precise floating point data interchange, not for printing out OCaml constants.

Very good point.

> Following established format OCaml convention for strings, char and floats, namely that uppercase format string means "OCaml syntax", I think we should rather only change the format for `%H` (hoping no-one is using it for interchange at the moment).

`%H` is now around for three years so this might indeed be rather risky.

Another solution mentioned in https://github.com/ocaml/ocaml/pull/268#issue-48880392 (sorry, should have read it before) is to use `%#F`. This seems better has it is currently unused and doesn't bring any backward compatibility issue. I'm updating the current PR according to this idea.

**Comment by proux01** [comment]
So this PR now offers to add `%#F` to output OCaml parsable floats in hexadecimal notations (just as `%h` except for -infinity and -nan).
`%F` is left unmodified and still outputs decimal notations.
Thus, the proposed change should be perfectly backward compatible.

**Comment by proux01** [comment]
Ping?

**Comment by damiendoligez** [comment]
cc @xavierleroy @gasche

File: stdlib/camlinternalFormat.ml
Comment (gasche) [inline stdlib/camlinternalFormat.ml]: I would prefer to avoid the fragile pattern here, by listing all the cases instead of using `_`. This will prevent a bug if we decide to add another `#..` float format later.

File: stdlib/camlinternalFormat.ml
Comment (gasche) [inline stdlib/camlinternalFormat.ml]: I would break the line before `begin` to clarify the two-steps structure.

File: stdlib/camlinternalFormat.ml
Comment (gasche) [inline stdlib/camlinternalFormat.ml]: The whole `match snd fconv with ...` match would be more readable if the logic of this big `let str` block was moved into a function definition at the same place as `let hex ()`. (I wondered about suggesting to move both to outside the function, but they would need better name and the current local placement is probably fine.) What about `maybe_add_dot` or `add_dot_if_needed` as a function name (`string -> string`)?

File: stdlib/camlinternalFormat.ml
Comment (gasche) [inline stdlib/camlinternalFormat.ml]: I would break the line before `(get_plus ())`, rather than before `(get_space ())`, the current placement (even before the patch) has no logic.

**Review by gasche — COMMENTED** [review]
Sorry for dropping the ball on this one...

I made a review for code correctness, and everything looks good (as I learned to expect from @proux01). I only made a few readability comments.

I'm not enough of a floating-point user to have a strong opinion on the desirability of the feature and the ergonomics of the choice of `%#F`. I would non-strongly approve of the choices that have been made -- I think using `#` for this makes as much sense as other (fairly ad-hoc) uses of `#`, and that the feature does sound useful.

I would be interested in a second opinion on this aspect (is it useful? is it the right interface?), which is why I haven't "Approved" yet.

File: stdlib/printf.mli
Comment (xavierleroy) [inline stdlib/printf.mli]: Instead of `or float ([F])`, I'd rather be more explicit: `and the floating-point type [F]`.

Although this predates this change, I don't understand why the enumeration `[x], ... [nu]` is between parentheses.

**Review by xavierleroy — COMMENTED** [review]
Like @gasche I don't feel strongly about this feature.  If we agree it is useful, the implementation looks good and the choice of format (`%#F`) reasonable, given that `%H` is taken already.

**Review by gasche — APPROVED** [review]
Then I propose to move forward and merge the PR once the review comments are addressed. @proux01, will you have time to deal with them? (Otherwise I could take care of them at some point.)

File: stdlib/camlinternalFormat.ml
Comment (proux01) [reply stdlib/camlinternalFormat.ml]: Indeed, done

File: stdlib/camlinternalFormat.ml
Comment (proux01) [reply stdlib/camlinternalFormat.ml]: Right, done

File: stdlib/camlinternalFormat.ml
Comment (proux01) [reply stdlib/camlinternalFormat.ml]: This does enlarge the diff but for the better. Done

File: stdlib/camlinternalFormat.ml
Comment (proux01) [reply stdlib/camlinternalFormat.ml]: done

File: stdlib/printf.mli
Comment (proux01) [reply stdlib/printf.mli]: > Instead of or float ([F]), I'd rather be more explicit: and the floating-point type [F].

Right, done

> Although this predates this change, I don't understand why the enumeration [x], ... [nu] is between parentheses.

I guess because it is an enumeration of *all* integer formats (`nx`, `nX` and `no` are missing but actually supported). Since this complete enumeration is quite useless, I rather offer to remove it and add a slightly more precise documentation on the bahavior of `#` on each integer format.

**Comment by proux01** [comment]
Thanks @gasche ! Comments adressed, let me know if a rebase is needed.

**Comment by gasche** [comment]
It would be nice if you could rebase the PR on top of `trunk` indeed, to have the CI run again. This should be good to go when the CI passes.

**Comment by proux01** [comment]
@gasche : rebased, CI green

**Pull request merged**
